internal/app/machined/pkg/controllers/siderolink/manager.go

```diff
@@ -196,10 +286,14 @@ func (ctrl *ManagerController) Run(ctx context.Context, r controller.Runtime, lo
 			return fmt.Errorf("error parsing node address: %w", err)
 		}
 
-		if err = r.Modify(ctx,
-			network.NewLinkSpec(network.ConfigNamespaceName, network.LayeredID(network.ConfigOperator, network.LinkID(constants.SideroLinkName))),
-			func(r resource.Resource) error {
-				spec := r.(*network.LinkSpec).TypedSpec()
+		if err = ctrl.cleanup(ctx, r, linkSpec.Metadata()); err != nil {
```

Comment:
here it's more of a nit, but it'd be nice if we could not delete the `siderolink` link and addresses if they don't change.

that might need some refactoring once again to make sure things work correctly. 

I would try to imagine this controller outputting resources: LinkSpec & AddresSpec, and a cleanup which should follow it - remove owned LinkSpec & AddressSpec which are no longer touched by the controller.

One case which doesn't work right now is that if `siderolink.Config` is dropped, this controller should tear down the link & address

Reply:
> I would try to imagine this controller outputting resources: LinkSpec & AddresSpec, and a cleanup which should follow it - remove owned LinkSpec & AddressSpec which are no longer touched by the controller.

👍 

> One case which doesn't work right now is that if `siderolink.Config` is dropped, this controller should tear down the link & address

Actually we do that: https://github.com/siderolabs/talos/pull/7175/files#diff-37ca5158dd35364ca3b988bd5c946b7cb8488d0a595212a6dbff527072de904cR216-R218

Tested as well: https://github.com/siderolabs/talos/pull/7175/files#diff-cd9530d76c3a8d8f60adfb9f54af2b6feff998e01388da405db06ccf92bc51aaR148-R170

Or do you mean something else?

Reply:
> I would try to imagine this controller outputting resources: LinkSpec & AddresSpec, and a cleanup which should follow it - remove owned LinkSpec & AddressSpec which are no longer touched by the controller.

This is also done now.